boofuzz/monitors/process_monitor.py

```diff
@@ -0,0 +1,71 @@
+from . import pedrpc
+from .imonitor import IMonitor
+
+
+# Important: IMonitor needs to come *before* pedrpc.Client in the
+# Inheritance list for the method resolution order to produce
+# correct results.
+
+
+class ProcessMonitor(IMonitor, pedrpc.Client):
+    """
+    Proxy class for the process monitor interface.
+
+    In Versions < 0.2.0, boofuzz had network and process monitors
+    that communicated over RPC. The RPC client was directly passed
+    to the session class, and resolved all method calls dynamically
+    on the RPC partner.
+
+    Since 0.2.0, every monitor class must implement the abstract class
+    IMonitor, which defines a common interface among all Monitors. To
+    aid future typehinting efforts and to make Network- and Process Monitors
+    disambiguable, this explicit proxy class has been introduced that
+    fast-forwards all calls to the RPC partner.
+
+    .. versionadded:: 0.2.0
+    """
+
+    def __init__(self, host, port):
+        IMonitor.__init__(self)
+        pedrpc.Client.__init__(self, host, port)
+
+    def alive(self):
+        return pedrpc.Client.alive(self)  # pytype: disable=attribute-error
```

Comment:
This still doesn't work as expected.
If we want calls to ProcessMonitor.alive() to be delegated to pedrpc.Client, we mustn't overwrite it here.
That again is a little problematic with `alive` being abstract in `IMonitor`. Not exactly sure how to solve this.

Reply:
But the method resolution order is ProcessMonitor -> IMonitor -> pedrpc.Client, so explicitely dispatching alive() should reach pedrpc.Client, not IMonitor. Not overriding the Method would, in any case, tingle down to IMonitor even if we define it as abstract.

Not sure how to proceed here, I guess I need to experiment with pythons inheritance a little more.

Reply:
Yes, I have tested this and ran into errors.

I didn't have time to look into this, but `pedrpc.Client.alive(self)` seems to be generating a new object instead of using the one we create in the constructor.
Completely removing `alive()` from IMonitor and ProcessMonitor worked just fine.

Reply:
I did have a quick glance but since it didn't error out on me, I gave it now further thought.

I have done more research on the whole inheritance thing, and I think that I can clear up some of the confusion now.

I still have no idea why `pedrpc.Client.alive(self)` does not work. For all intents and purposes, this should behave like a static method call supplying an explicit self. As long it is type-compatible (which it should be because, well, polymorphism), that should work. It could be that this is an edge case of the multiple inheritance I am not aware of, however I am fairly certain that it does not generate a new object. If you could share your testing code, that would be great, because by now I am really interested in what is going on and can't figure it out from documentation and your description.

The good news is that I finally have found a way to make things work, but it comes with a catch. The problem with my solution is that multiple inheritance is conceptually not the right way to do things here, and this is why this is confusing and complicated. @SR4ven you were right that I got confused about supers arguments. super(B, self) with a MRO of, say, B -> C -> A only searches C and A for matching members; super(C, self) in the same MRO only searches A as it skips the provided type. Which is a reasonable behavior, but if you're used to statically typed languages it's easy to get confused because it's usually inclusive there (like `((C)this).foobar()` searches C and A). However, super() just walks up the MRO, so sibling ancestors are considered as well (yay, more confusion!).

To fix this once and for all, there are two possibilities:

- Option A: add a shim class between pedrpc.Client and the (Network|Process)Monitor classes and use this to massage super() into skipping the IMonitor part of the MRO. Pro: Easy to implement, relatively easy to understand if the shim class is suffiently documentated with its reasoning. Con: Makes the inheritance tree messy and is not a general case solution.
- Option B: Modify IMonitor to try to call super(IMonitor, self) in each method and only executing its own body (if any) if super() fails with an attribute error.  Pro: should solve the problem in general, even if it would be encountered in different form. Con: Makes the IMonitor even messier than it is.

For Option B, the super-shimming could eventually be implemented as a decorator making it a little prettier. At least, this would be the cooperative-multiple-inheritance-way of doing things.

Or we could just drop IMonitor from the inheritance list and manually worry about the methods that should exist but not be RPC forwarded. Not really the OOP way of doing things but pragmatic, given the mess that would be introduced with either the solutions that keep the multiple inheritance. I'm still unsure, however I will make a minimal working example of the inheritance problem to tinker around with.

Reply:
Thanks for the detailed explanation @mistressofjellyfish!

I ran the test with a Session object like this:
```python
sess = Session(
    target=Target(connection=SocketConnection("127.0.0.1", 80, proto="tcp"),
                  monitors=[ProcessMonitor("localhost", 26002)]),
```
(If that's not how it's supposed to be used we might want to add an example to the docs)

```
Traceback (most recent call last):
  File "/home/dev/Documents/boofuzz_requests/dev.py", line 110, in <module>
    fuzz()
  File "/home/dev/Documents/boofuzz_requests/dev.py", line 91, in fuzz
    sleep_time=1,
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/sessions.py", line 539, in __init__
    self.add_target(target)
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/sessions.py", line 577, in add_target
    target.monitors_alive()
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/sessions.py", line 147, in monitors_alive
    if monitor.alive():
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/monitors/process_monitor.py", line 33, in alive
    return pedrpc.Client.alive(self)  # pytype: disable=attribute-error
AttributeError: type object 'Client' has no attribute 'alive'
```

Reply:
Now that I look at it, I know exactly why this doesn't work, and it has exactly to do with why this is such a mess in the first place. The functions in the class get dispatched dynamically by __method_missing, which gets called by __getattr__, which normaly get's called in the process of doing a selfcall. But because we call statically providing an explicit self, of course __getattr__ never gets called.

So, in this case, if we don't call pedrpc.Client.alive but instead pedrpc.Client.__method_missing(self, "alive") it should work. Man this is a mess.

Reply:
So I finally can resolve this. What's in place now isn't the most clean solution as it calls __method_missing explicitely, but I didn't find any other way without re-thinking the whole RPC interface.

Oh well, it works now. If I was to design that from scratch I'd build something around RPyC or whatever.